### PR TITLE
Initial commit of idol-nifi chart

### PR DESCRIPTION
**Pull request by daniel-lau-mf**

### CODE REVIEW

File: helm/idol-nifi/values.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# idolImageRegistry - used to construct container image name: <idolImageRegistry>/<nifiImage>:<idolVersion>
+idolImageRegistry: microfocusidolserver
+# idolVersion - used to construct container image name: <idolImageRegistry>/<nifiImage>:<idolVersion>
+idolVersion: "23.4"
+
+# imagePullSecrets - secrets used to pull container images
+imagePullSecrets:
+- "dockerhub-secret"
+
+nifi:
```
Comment (christopher-rayson-mf) [inline helm/idol-nifi/values.yaml:17]: Do we have a good reason to use nested over flat here (cf. [Best Practices](https://helm.sh/docs/chart_best_practices/values/#flat-or-nested-values), _"In most cases, flat should be favored over nested."_)?

TBH I think I prefer the structure of nested, but we deliberately flattened (most) values in distributed-idol to comply with the above advice.

File: helm/idol-nifi/templates/nifi/service.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  labels: {{- include "idol-nifi.labels" . | nindent 4 -}}
+    app: nifi
+  annotations:
+    app.kubernetes.io/name: nifi
+    app.kubernetes.io/part-of: nifi
+spec:
+  type: ClusterIP
+  selector:
+    app: nifi
+  ports:
```
Comment (christopher-rayson-mf) [inline helm/idol-nifi/templates/nifi/service.yaml:15]: Is NiFi going to expose an ACI port (for Universal Viewing)?

File: helm/idol-nifi/templates/nifi/scripts-configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: nifi-ssl-cm
+  labels: {{- include "idol-nifi.labels" . | nindent 4 -}}
+    app: nifi
+data:
+  security.sh: |
```
Comment (christopher-rayson-mf) [inline helm/idol-nifi/templates/nifi/scripts-configmap.yaml:8]: Easier to manage/edit if scripts are checked in separately and transcluded by `.Files.Get`?

File: helm/idol-nifi/values.yaml (hunk 1 above)
Comment (daniel-lau-mf) [reply helm/idol-nifi/values.yaml:17]: > When there are a large number of related variables, and at least one of them is non-optional, nested values may be used to improve readability.

I much prefer structure too and figurem can justify readability winning out. 

> For optimal safety, a nested value must be checked at every level
I think mostly this is mitigated by having default values, which ensures the structure?

Comment (christopher-rayson-mf) [reply helm/idol-nifi/values.yaml:17]: Fair, that's a given exception to the rule. ✔️

File: helm/idol-nifi/templates/nifi/nifi.yaml [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+              drop:
+                - ALL
+      imagePullSecrets:
+        {{- range .Values.imagePullSecrets }}
+        - name: {{ . }}
+        {{- end }}
+      containers:
+      - name: nifi
+        image: {{ .Values.idolImageRegistry }}/{{ .Values.nifi.imageName }}:{{ .Values.idolVersion }}
+        #image: idol-docker-images.svsartifactory.swinfra.net/nifi-full:23.4
```
Comment (christopher-rayson-mf) [inline helm/idol-nifi/templates/nifi/nifi.yaml:73]: Intended to commit this line?

File: helm/idol-nifi/templates/nifi/nifi.yaml [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                        cluster_node_status=`curl -s http://$host:$port/nifi-api/controller/cluster/nodes/$cluster_node_id | jq .node.status -r`
+                    done
+                }
+
+                if [ ! -z "$cluster_node_id" ]
+                then
+                    setAndWaitForStatus "DISCONNECTING" "DISCONNECTED"
+                    setAndWaitForStatus "OFFLOADING" "OFFLOADED"
+                    curl -s -X DELETE http://$host:$port/nifi-api/controller/cluster/nodes/$cluster_node_id
+                fi
```
Comment (christopher-rayson-mf) [inline helm/idol-nifi/templates/nifi/nifi.yaml:164]: Worth putting this into a separate .sh file in the `scripts` mount?

File: helm/idol-nifi/templates/nifi-registry/configmap.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: nifi-registry-cm
+  labels: {{- include "idol-nifi.labels" . | nindent 4 -}}
+    app: nifi-registry
+data:
+
```
Comment (christopher-rayson-mf) [inline helm/idol-nifi/templates/nifi-registry/configmap.yaml:8]: Intentionally blank?

File: helm/idol-nifi/templates/nifi/nifi.yaml [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+          timeoutSeconds: 10
+          failureThreshold: 3
+          successThreshold: 1
+        resources:
+          requests:
+            cpu: 400m
+            memory: 2Gi
+          limits:
+            cpu: 400m
+            memory: 2Gi
```
Comment (daniel-lau-mf) [inline helm/idol-nifi/templates/nifi/nifi.yaml:189]: These should be configurable and probably higher default values.
See also NIFI_JVM_HEAP_INIT/NIFI_JVM_HEAP_MAX in the `configmap.yaml`

File: helm/idol-nifi/templates/nifi/secrets.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+type: Opaque
+metadata:
+  name: nifi-basic-auth
+  labels: {{- include "idol-nifi.labels" . | nindent 4 }}
+    app: nifi
+  annotations:
+    app.kubernetes.io/name: nifi
+    app.kubernetes.io/part-of: nifi
+data:
+  auth: YWRtaW46JGFwcjEkSDY1dnBkTU8kMXAxOGMxN3BuZVFUT2ZjVC9TZkZzMQo=
```
[the remark this answers is not in the document]
Comment (daniel-lau-mf) [reply helm/idol-nifi/templates/nifi/secrets.yaml:13]: It's a hard-coded admin/admin. Can try to make this more configurable.

[the remark this answers is not in the document]
Comment (christopher-rayson-mf) [reply helm/idol-nifi/templates/nifi/secrets.yaml:13]: Find is also hardcoded Admin/Admin...

File: helm/idol-nifi/templates/nifi/service.yaml (hunk 2 above)
Comment (daniel-lau-mf) [reply helm/idol-nifi/templates/nifi/service.yaml:15]: [e4dbdbe](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/e4dbdbe5ac4d8d45fbb1e33020f6c9e5f5d034be)

File: helm/idol-nifi/templates/nifi/secrets.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (daniel-lau-mf) [reply helm/idol-nifi/templates/nifi/secrets.yaml]: [387aae9](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/387aae96a336cc70781981821a14979dae97468c)

File: helm/idol-nifi/templates/nifi/scripts-configmap.yaml (hunk 3 above)
Comment (daniel-lau-mf) [reply helm/idol-nifi/templates/nifi/scripts-configmap.yaml:8]: [8f12d06](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/8f12d06cfce23e46b00d89a6623ee25d4c4664c3)

File: helm/idol-nifi/templates/nifi/nifi.yaml (hunk 4 above)
Comment (daniel-lau-mf) [reply helm/idol-nifi/templates/nifi/nifi.yaml]: [bdcf610](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/bdcf6102a5b4a6abe48f8d6a95c252e1e6a5c6c4)

File: helm/idol-nifi/templates/nifi/nifi.yaml (hunk 5 above)
Comment (daniel-lau-mf) [reply helm/idol-nifi/templates/nifi/nifi.yaml]: [8f12d06](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/8f12d06cfce23e46b00d89a6623ee25d4c4664c3)

File: helm/idol-nifi/templates/nifi-registry/configmap.yaml (hunk 6 above)
Comment (daniel-lau-mf) [reply helm/idol-nifi/templates/nifi-registry/configmap.yaml:9]: [bdcf610](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/bdcf6102a5b4a6abe48f8d6a95c252e1e6a5c6c4)

File: helm/idol-nifi/templates/nifi/nifi.yaml (hunk 7 above)
Comment (daniel-lau-mf) [reply helm/idol-nifi/templates/nifi/nifi.yaml]: [438faf0](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/438faf01f6eb7f7d02fe5f2c5ae42a0874f81af8)

File: helm/idol-nifi/resources/flow-basic-idol.json [hunk 9]
```diff
@@ -1 +1 @@
...
...false,"name":"AllowMultipleResults","sensitive":false},"[PIIValidation]Entities":{"displayName":"[PIIValidation]Entities","identifiesControllerService":false,"name":"[PIIValidation]Entities","sensitive":false}},"retriedRelationships":[],"retryCount":10,"runDurationMillis":0,"scheduledState":"ENABLED","schedulingPeriod":"0 sec","schedulingStrategy":"TIMER_DRIVEN","style":{},"type":"idol.nifi.processor.Eduction","yieldDuration":"1 sec"}],"remoteProcessGroups":[],"variables":{}},"flowEncodingVersion":"1.0","parameterContexts":{"IDOL NiFi Flow Parameters":{"componentType":"PARAMETER_CONTEXT","inheritedParameterContexts":[],"name":"IDOL NiFi Flow Parameters","parameters":[{"description":"","name":"IndexHost","provided":false,"sensitive":false,"value":"idol-index-service"},{"description":"","name":"LicenseServerHost","provided":false,"sensitive":false,"value":"idol-licenseserver"},{"description":"","name":"IndexACIPort","provided":false,"sensitive":false,"value":"9070"},{"description":"","name":"LicenseServerACIPort","provided":false,"sensitive":false,"value":"20000"}]}},"parameterProviders":{},"snapshotMetadata":{"author":"anonymous","comments":"","timestamp":1699008213640,"version":2}}
```
Comment (christopher-rayson-mf) [inline helm/idol-nifi/resources/flow-basic-idol.json:1]: In the parameterized file, there's an "IdolLicenseServiceImpl" with
```json
"properties": {
    "LicenseServerHost": "#{LicenseServerHost}",
    "LicenseServerPort": "#{LicenseServerACIPort}"
}
```
but also a "FileSystem Connector Group IdolLicenseService" with
```json
"properties": {
    "LicenseServerHost": "${license.server.hostname}",
    "LicenseServerPort": "20000"
}
```
Guessing that last port should also be parameterized?

Comment (christopher-rayson-mf) [reply helm/idol-nifi/resources/flow-basic-idol.json:1]: Just noticed the format of the hostname is different too.

Comment (daniel-lau-mf) [reply helm/idol-nifi/resources/flow-basic-idol.json]: [41c794d](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/41c794d8bdd5e7cad9638655c7111306f1baf547)

Comment (christopher-rayson-mf) [reply helm/idol-nifi/resources/flow-basic-idol.json]: 👍 

The FileSystem Connector Group has a now-redundant `license.server.hostname` variable; don't know whether this shows up in any way.

Comment (daniel-lau-mf) [reply helm/idol-nifi/resources/flow-basic-idol.json]: [25ce39a](https://github.com/opentext-idol/idol-containers-toolkit/pull/12/commits/25ce39ace958b3b7020bbfc0ba36accb5a79cb20)

**Comment by christopher-rayson-mf** [comment]
## `logs/post-start.log`
```
[Fri Nov 3 17:08:56 UTC 2023] Got registry client: 1   nifi-registry   962881af-018b-1000-0000-00001a605e52   http://nifi-registry:18080
Checking for flow IDOL NiFi Flow
RC=1 OUTPUT=
BUCKETID=0ea1fb40-a9d3-4300-8a9d-1d4b81019fb7
FLOWID=ba12fb2d-641e-4f04-9461-0897cbca7506
FLOWVERSION=1
PROCESSGROUP=
ERROR: Error executing command 'pg-import' : Error creating process group: java.net.SocketTimeoutException: timeout
PARAMCONTEXT=
ERROR: Error executing command 'pg-get-param-context' : Error getting process group: The request was rejected because the URL contained a potentially malicious String "%0A"

ERROR: Error executing command 'set-param' : Error retrieving parameter context: Unable to find Parameter Context with id 'ERROR:'.


ERROR: Error executing command 'set-param' : Error retrieving parameter context: Unable to find Parameter Context with id 'ERROR:'.


ERROR: Error executing command 'set-param' : Error retrieving parameter context: Unable to find Parameter Context with id 'ERROR:'.


ERROR: Error executing command 'set-param' : Error retrieving parameter context: Unable to find Parameter Context with id 'ERROR:'.

Enabling services


ERROR: Error executing command 'pg-enable-services' : Error retrieving controller services: The request was rejected because the URL contained a potentially malicious String "%0A"


org.apache.nifi.toolkit.cli.api.CommandException: Error executing command 'pg-enable-services' : Error retrieving controller services: The request was rejected because the URL contained a potentially malicious String "%0A"
        at org.apache.nifi.toolkit.cli.impl.command.nifi.AbstractNiFiCommand.doExecute(AbstractNiFiCommand.java:65)
        at org.apache.nifi.toolkit.cli.impl.command.AbstractPropertyCommand.execute(AbstractPropertyCommand.java:74)
        at org.apache.nifi.toolkit.cli.impl.command.CommandProcessor.processCommand(CommandProcessor.java:252)
        at org.apache.nifi.toolkit.cli.impl.command.CommandProcessor.processGroupCommand(CommandProcessor.java:233)
        at org.apache.nifi.toolkit.cli.impl.command.CommandProcessor.process(CommandProcessor.java:188)
        at org.apache.nifi.toolkit.cli.CLIMain.runSingleCommand(CLIMain.java:145)
        at org.apache.nifi.toolkit.cli.CLIMain.main(CLIMain.java:72)
Caused by: org.apache.nifi.toolkit.cli.impl.client.nifi.NiFiClientException: Error retrieving controller services: The request was rejected because the URL contained a potentially malicious String "%0A"
        at org.apache.nifi.toolkit.cli.impl.client.nifi.impl.AbstractJerseyClient.executeAction(AbstractJerseyClient.java:90)
        at org.apache.nifi.toolkit.cli.impl.client.nifi.impl.JerseyFlowClient.getControllerServices(JerseyFlowClient.java:194)
        at org.apache.nifi.toolkit.cli.impl.command.nifi.pg.cs.ControllerServiceUtil.getControllerServiceStates(ControllerServiceUtil.java:33)
        at org.apache.nifi.toolkit.cli.impl.command.nifi.pg.PGEnableControllerServices.getControllerServiceStates(PGEnableControllerServices.java:163)
        at org.apache.nifi.toolkit.cli.impl.command.nifi.pg.PGEnableControllerServices.doExecute(PGEnableControllerServices.java:81)
        at org.apache.nifi.toolkit.cli.impl.command.nifi.pg.PGEnableControllerServices.doExecute(PGEnableControllerServices.java:40)
        at org.apache.nifi.toolkit.cli.impl.command.nifi.AbstractNiFiCommand.doExecute(AbstractNiFiCommand.java:63)
        ... 6 more
Caused by: javax.ws.rs.InternalServerErrorException: HTTP 500 Server Error
        at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:960)
        at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:755)
        at org.glassfish.jersey.client.JerseyInvocation.lambda$invoke$1(JerseyInvocation.java:675)
        at org.glassfish.jersey.client.JerseyInvocation.call(JerseyInvocation.java:697)
        at org.glassfish.jersey.client.JerseyInvocation.lambda$runInScope$3(JerseyInvocation.java:691)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:205)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:390)
        at org.glassfish.jersey.client.JerseyInvocation.runInScope(JerseyInvocation.java:691)
        at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:674)
        at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:422)
        at org.glassfish.jersey.client.JerseyInvocation$Builder.get(JerseyInvocation.java:318)
        at org.apache.nifi.toolkit.cli.impl.client.nifi.impl.JerseyFlowClient.lambda$getControllerServices$5(JerseyFlowClient.java:199)
        at org.apache.nifi.toolkit.cli.impl.client.nifi.impl.AbstractJerseyClient.executeAction(AbstractJerseyClient.java:76)
        ... 12 more

Starting processors

ERROR: Error executing command 'pg-start' : Error scheduling components: The request was rejected because the URL contained a potentially malicious String "%0A"


org.apache.nifi.toolkit.cli.api.CommandException: Error executing command 'pg-start' : Error scheduling components: The request was rejected because the URL contained a potentially malicious String "%0A"
        at org.apache.nifi.toolkit.cli.impl.command.nifi.AbstractNiFiCommand.doExecute(AbstractNiFiCommand.java:65)
        at org.apache.nifi.toolkit.cli.impl.command.AbstractPropertyCommand.execute(AbstractPropertyCommand.java:74)
        at org.apache.nifi.toolkit.cli.impl.command.CommandProcessor.processCommand(CommandProcessor.java:252)
        at org.apache.nifi.toolkit.cli.impl.command.CommandProcessor.processGroupCommand(CommandProcessor.java:233)
        at org.apache.nifi.toolkit.cli.impl.command.CommandProcessor.process(CommandProcessor.java:188)
        at org.apache.nifi.toolkit.cli.CLIMain.runSingleCommand(CLIMain.java:145)
        at org.apache.nifi.toolkit.cli.CLIMain.main(CLIMain.java:72)
Caused by: org.apache.nifi.toolkit.cli.impl.client.nifi.NiFiClientException: Error scheduling components: The request was rejected because the URL contained a potentially malicious String "%0A"        at org.apache.nifi.toolkit.cli.impl.client.nifi.impl.AbstractJerseyClient.executeAction(AbstractJerseyClient.java:90)
        at org.apache.nifi.toolkit.cli.impl.command.nifi.AbstractNiFiCommand.doExecute(AbstractNiFiCommand.java:63)
        ... 6 more
Caused by: javax.ws.rs.InternalServerErrorException: HTTP 500 Server Error
        at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:960)
        at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:755)
        at org.glassfish.jersey.client.JerseyInvocation.lambda$invoke$1(JerseyInvocation.java:675)
        at org.glassfish.jersey.client.JerseyInvocation.call(JerseyInvocation.java:697)
        at org.glassfish.jersey.client.JerseyInvocation.lambda$runInScope$3(JerseyInvocation.java:691)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:205)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:390)
        at org.glassfish.jersey.client.JerseyInvocation.runInScope(JerseyInvocation.java:691)
        at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:674)
        at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:450)
        at org.glassfish.jersey.client.JerseyInvocation$Builder.put(JerseyInvocation.java:334)
        at org.apache.nifi.toolkit.cli.impl.client.nifi.impl.JerseyFlowClient.lambda$scheduleProcessGroupComponents$3(JerseyFlowClient.java:155)
        at org.apache.nifi.toolkit.cli.impl.client.nifi.impl.AbstractJerseyClient.executeAction(AbstractJerseyClient.java:76)
        ... 10 more
```

**Review by christopher-rayson-mf — APPROVED** [review]
Question as to whether security context objects should be `required` in the schema; as they're present in the default values YAML, users would have to go well out their way to break them, so minor niggle at worst,

Other than that looks fine.

Attempt to deploy locally appeared to timeout executing `pg-import` during initialization (see log snippet in comments), but eventually started up with a recoverable flow.

**Review by christopher-rayson-mf — APPROVED** [review]

**Pull request merged**